src/services/db/GitFileSystemService.ts

```diff
@@ -623,17 +631,177 @@ export default class GitFileSystemService {
+            if (error instanceof NotFoundError) {
+              return okAsync(true)
+            }
+
+            return errAsync(error)
+          })
+      )
+      .andThen(() =>
+        ResultAsync.fromPromise(
+          this.git.cwd(`${EFS_VOL_PATH}/${repoName}`).mv(oldPath, newPath),
+          (error) => {
+            logger.error(`Error when moving ${oldPath} to ${newPath}: ${error}`)
+
+            if (error instanceof GitError) {
+              return new GitFileSystemNeedsRollbackError(
+                `Unable to rename ${oldPath} to ${newPath}`
+              )
+            }
+
+            return new GitFileSystemNeedsRollbackError(
+              "An unknown error occurred"
+            )
+          }
+        )
+      )
+      .andThen(() =>
+        this.commit(
+          repoName,
+          [oldPath, newPath],
+          userId,
+          message || `Renamed ${oldPath} to ${newPath}`
+        )
+      )
+      .orElse((error) => {
+        if (error instanceof GitFileSystemNeedsRollbackError) {
+          return this.rollback(repoName, oldStateSha).andThen(() =>
+            errAsync(new GitFileSystemError(error.message))
+          )
+        }
+
+        return errAsync(error)
+      })
+  }
+
+  // Move multiple files from oldPath to newPath without renaming them
+  moveFiles(
+    repoName: string,
+    oldPath: string,
+    newPath: string,
+    userId: string,
+    targetFiles: string[],
+    message?: string
+  ): ResultAsync<string, GitFileSystemError | ConflictError | NotFoundError> {
+    let oldStateSha = ""
+
+    return this.getLatestCommitOfBranch(repoName, BRANCH_REF)
+      .andThen((latestCommit) => {
+        // It is guaranteed that the latest commit contains the SHA hash
+        oldStateSha = latestCommit.sha as string
+        return okAsync(true)
+      })
+      .andThen(() => this.getFilePathStats(repoName, oldPath))
+      .andThen(() =>
+        // Ensure that the new path exists
+        ResultAsync.fromPromise(
+          fs.promises.mkdir(`${EFS_VOL_PATH}/${repoName}/${newPath}`, {
+            recursive: true,
+          }),
+          (error) => {
+            logger.error(`Error when creating ${newPath} during move: ${error}`)
+
+            if (error instanceof Error) {
+              return new GitFileSystemNeedsRollbackError(
+                `Unable to create ${newPath}`
+              )
+            }
+
+            return new GitFileSystemNeedsRollbackError(
+              "An unknown error occurred"
+            )
+          }
+        )
+      )
+      .andThen(() =>
+        combine(
+          targetFiles.map((targetFile) =>
+            // We expect to see an error here, since the new path should not exist
+            this.getFilePathStats(repoName, `${newPath}/${targetFile}`)
+              .andThen(() =>
+                errAsync(new ConflictError("File path already exists"))
+              )
+              .map(() => true)
+              .orElse((error) => {
```

Comment:
The whole operation will be aborted, because if any of these throw an error, the `combine()` will be errored, so we will initiate the rollback operation.

Reply:
got it